aws/validators.go

```diff
@@ -117,15 +117,19 @@ func validateTransferServerID(v interface{}, k string) (ws []string, errors []er
 
 func validateTransferUserName(v interface{}, k string) (ws []string, errors []error) {
 	value := v.(string)
-
 	// https://docs.aws.amazon.com/transfer/latest/userguide/API_CreateUser.html
-	pattern := `^[a-z0-9]{3,32}$`
-	if !regexp.MustCompile(pattern).MatchString(value) {
-		errors = append(errors, fmt.Errorf(
-			"%q isn't a valid transfer user name (only lowercase alphanumeric characters are allowed): %q",
-			k, value))
+	if regexp.MustCompile(`[^0-9a-zA-Z_-]`).MatchString(value) {
+		errors = append(errors, fmt.Errorf("%q can only contain alphanumeric characters, underscores, and hyphens", k))
+	}
+	if len(value) < 3 {
```

Comment:
I would start by checking the length first.
Then the patterns: maybe "cannot begin with a hyphen" first?

Reply:
I'll update it now, does this order sound good?

Start with hyphen
Less than 3 characters
More than 32 characters
Regex match

Reply:
I believe that if the length is not good there is no point checking for characters so I'd start with it.  
Otherwise, you might just use [strings.HasPrefix](https://golang.org/pkg/strings/#HasPrefix) for the "start with hypen" rule.

Reply:
Would you rather have a single conditional? I was only concerned about the longer output but if a single is preferred I'll keep that in.